### PR TITLE
[ENHANCEMENT] Adequations on the project for `Heroku` deployment

### LINKED ISSUE #9: Publish the service on Heroku as a POC

### LINKED ISSUE #10: Create a docker container with the sucuri-api REST API service

### PR DESCRIPTION
**Pull request by Mxrcon**
Hey @abhi18av and @Aryan-sharma11, I took the day to hard focus on this project and try to have a successful deployment, but I faced some problems into the containerization and I'm here to explain how I got over them.

1. I had to change `Get-Version` so it works with any available conda binary, giving our module a lot more of generalization and hopefully more portable to different infra
2. As Heroku doesn't have native support to Powershell projects, I decided to deploy there using a container with the following strucuture:
```terminal
# on sucuri-api/containers
heroku container:push web --app sucuri-api
heroku container:release web --app sucuri-api
```
3. For a complete deployment some Docker images changes were necessary, like `EXPOSE $PORT` to use heroku port
4. I found a bug, that can be further explored on future: If conda doesn't have any channel avaible, the API will always throw 404

This api is avaible at `sucuri-api.herokuapp.com`, Please give it a try using your favorite api call method. 

I'm completely happy with the overall development, the initial setup from @Aryan-sharma11 was already amazing, and now we can access it with even a custom domain.  

This Pull-Request closes #9 and closes #10. Also is a big step at our roadmap!

I kindly request your review as this are Major changes to the docker container which I need to admit that is my weakest point on dev-ops.

### CODE REVIEW

**Comment by abhi18av** [comment]
This is great progress @Mxrcon!

From this real world experience, I think we can take up the following tasks (which I'll expand in specific tasks)

- [ ] Allow multiple-channels to be specified in the request as query parameters or json data
- [ ] Implement a default message on the base API URL `sucuri.../`

File: Sucuri-API/Private/FindCondaBinary.ps1 [hunk 1]
```diff
@@ -0,0 +1,23 @@
+<# Create a private function focused only on finding the conda-compatible binary #>
+function FindCondaBinary {
+    # find the best Conda binary avaible:
+    $condaBinary = ""
+    $possibleCondaBinaries = @("micromamba", "mamba", "conda")
+
+    #TODO: This can be done in parallel via ForEach-Object -Parallel option
```
Comment (abhi18av) [inline Sucuri-API/Private/FindCondaBinary.ps1:7]: Created a private function, but we are not using this as of now.

I don't know how to source a private function 😅

File: Sucuri-API/Public/Get-Version.ps1 [hunk 2]
```diff
@@ -12,11 +13,21 @@
...
+    # find the best Conda binary avaible:
+    $conda_bin = ""
+    $conda_binaries = @("micromamba","mamba","conda")
+    foreach($binarie in $conda_binaries){
+        if([bool](Get-Command -Name $binarie -ErrorAction SilentlyContinue)){
+	    $conda_bin = $binarie
+            break
+	    }
+	}
+    if($conda_bin -eq ""){throw "There's no Conda binary avaible."}
```
Comment (abhi18av) [inline Sucuri-API/Public/Get-Version.ps1:26]: This functionality should ideally be extracted into a private function 

CC @Aryan-sharma11

File: Sucuri-API/Public/Start-SucuriServer.ps1 [hunk 3]
```diff
@@ -29,32 +29,30 @@ function Start-SucuriServer {
...
 
-	# import Sucuri Module into the server
-	Import-Module -Name (Join-Path -Path (Join-Path (Split-Path -Path $PSScriptRoot -Parent) -ChildPath "") `
+      # import Sucuri Module into the server
+      Import-Module -Name (Join-Path -Path (Join-Path (Split-Path -Path $PSScriptRoot -Parent) -ChildPath "") `
             -ChildPath "Sucuri-API.psm1") 
 
-        Add-PodeEndpoint -Address localhost -Port 8085 -Protocol Http
-        Add-PodeRoute -Method 'Get' -Path '/' -ScriptBlock {
+      Add-PodeEndpoint -Address "0.0.0.0" -Port 8085 -Protocol Http
```
Comment (abhi18av) [inline Sucuri-API/Public/Start-SucuriServer.ps1:41]: @Mxrcon @Aryan-sharma11 , please use the VsCode powershell plugin formatting to maintain consistency in the codebase.

File: container/build.ps1 [hunk 4]
```diff
@@ -1,4 +0,0 @@
-# Command used to download the restish locally
-#aria2c "https://github.com/danielgtaylor/restish/releases/download/v0.14.0/restish-0.14.0-linux-x86_64.tar.gz"
-
-docker build -t sucuri-api:latest .
```
Comment (abhi18av) [inline container/build.ps1:4]: I've deleted this file and moved the content to the `psakefile`

File: container/Dockerfile [hunk 5]
```diff
@@ -34,22 +30,33 @@ ENTRYPOINT ["/usr/local/bin/_entrypoint.sh"]
...
 # expose the port
-EXPOSE 8085
+EXPOSE $PORT
+
+# Run this from project root 
+# Invoke-psake -buildFile ./psakeFile.ps1 -TaskList DockerBuild
+COPY . /home/sucuri-api/
+
+# clone the git repo
+#CMD [ "pwsh","-c","/opt/conda/bin/git", "clone --branch mxrcon/dev-heroku-deployment https://github.com/biosharp-dotnet/sucuri-api" ,"/home/sucuri-api;"]
```
Comment (abhi18av) [inline container/Dockerfile:58]: The automations should not rely upon a particular branch only, instead they should be usable from any branch.

Therefore, I've implemented a source copy mechanism, which puts the entire app within the container and then invokes powershell.

File: Sucuri-API/Public/Start-SucuriServer.ps1 [hunk 6]
```diff
@@ -29,32 +29,30 @@ function Start-SucuriServer {
...
+      Add-PodeRoute -Method 'Get' -Path '/' -ScriptBlock {
 
-        $pkg_name = $WebEvent.Query['pkg_name']
+         $pkg_name = $WebEvent.Query['pkg_name']
 
-        $response =  Get-version $pkg_name
+         $response = Get-version $pkg_name
 
-        if($response -ne "false"){
+         if ($response -match '\d\.\d\.\d|\d\.\d') {
```
Comment (abhi18av) [inline Sucuri-API/Public/Start-SucuriServer.ps1:48]: @Mxrcon , could you please add comments explaining why this was done?

**Review by abhi18av — COMMENTED** [review]
@Mxrcon @Aryan-sharma11 , I have done my code review now and have added some comments to explain the changes.

I've also requested some minor tweaks, but those can be covered in another PR (such as commenting code).

Approving the merge on this one now - but let's merge it after @Mxrcon rebuilds the container and deploys the new API version.

File: Sucuri-API/Private/FindCondaBinary.ps1 (hunk 1 above)
Comment (Mxrcon) [reply Sucuri-API/Private/FindCondaBinary.ps1:7]: :face_in_clouds: Just added that on the latest commit, good idea to use as private

File: Sucuri-API/Public/Get-Version.ps1 (hunk 2 above)
Comment (Mxrcon) [reply Sucuri-API/Public/Get-Version.ps1:26]: resolving as you already did that

File: Sucuri-API/Public/Start-SucuriServer.ps1 (hunk 3 above)
Comment (Mxrcon) [reply Sucuri-API/Public/Start-SucuriServer.ps1:41]: ok ok

File: Sucuri-API/Public/Start-SucuriServer.ps1 (hunk 6 above)
Comment (Mxrcon) [reply Sucuri-API/Public/Start-SucuriServer.ps1:48]: yes, this was done to cover the issue when there's no available conda channel to seek packages, and this also covers any other possible scenario when Get-Version doesn't return a valid version.

File: container/Dockerfile (hunk 5 above)
Comment (Mxrcon) [reply container/Dockerfile:58]: hmm, I understand, Thank you abhinav, this is a very good addition

**Comment by abhi18av** [comment]
Going ahead with the merge as the core POC has been completed and now we can finalize the overall details for long term maintenance and automations.

**Pull request merged**